File: backend/src/nausicass_global_green_initiative_api/services/monitoring/sentry_transaction.py
from typing import Any

import sentry_sdk

from .monitoring_interface import MonitoringTransaction


# Wrapper around Sentry's transaction class
class SentryTransaction(MonitoringTransaction):
    def __init__(self, name: str, op: str, **kwargs: Any) -> None:
        self._span = sentry_sdk.start_span(name=name, op=op)
        for key, value in kwargs.items():
            self._span.set_data(key, value)

    def finish(self) -> None:
        self._span.finish()

    def set_status(self, status: str) -> None:
        self._span.set_status(status)

File: frontend/src/services/monitoring/sentry/sentryTransaction.ts
import * as Sentry from "@sentry/react"

import type { MonitoringTransaction, TransactionContext } from "../types"

// Wrapper around Sentry's transaction to align with MonitoringTransaction interface
export class SentryTransaction implements MonitoringTransaction {
  // Reference to the underlying Sentry transaction
  private span: Sentry.Span | undefined

  constructor(context: TransactionContext) {
    // Updates: Sentry's startInactiveSpan is used to
    // create a transaction that can be started and finished
    // manually
    this.span = Sentry.startInactiveSpan({
      name: context.name,
      op: context.op,
      forceTransaction: true,
      ...(context.data && { attributes: context.data }),
    })
  }

  // Method to finish the transaction
  finish(): void {
    this.span?.end()
  }

  // Method to set the status of the transaction
  setStatus(status: string): void {
    this.span?.setStatus({ code: status === "ok" ? 1 : 2, message: status })
  }
}
